### PR TITLE
Deprecate Request and Response in Container

### PR DESCRIPTION
**Pull request by akrabat**
This PR adds deprecation notices when you retrieve the Request, Response or Environment from the Container. It's very confusing to have these in the Container as they are "stale" and should never be used.

It updates `App::run()` so that it no longer uses the container for these objects, but instead creates them itself. I've also updated `run()`'s signature so that you can now pass in a request and a response object if you want to use your own.

Similarly, I've updated `App::subRequest` to create it's own Environment and Response objects.

Would appreciate people's thoughts on this.

### CODE REVIEW

**Comment by mathmarques** [comment]
:+1: 
Would this not be a BC break?
If someone changes the container request and response object before calling run(), they would expect that Slim uses these objects, but instead, Slim will create a new request and response.

**Comment by akrabat** [comment]
That's a good question. They would have to change their code to send the request and response in via `run()`.

Maybe we need a new method then.

**Comment by mathmarques** [comment]
I guess everything should work now! 
Good job :+1:

**Comment by mnapoli** [comment]
I'm [still](https://github.com/slimphp/Slim/pull/1179#issuecomment-94328618) definitely :+1: on that ;) Those are value objects, not stateless services, so IMO they shouldn't be in the container.

**Comment by akrabat** [comment]
@mnapoli I'm annoyed with myself for not thinking of changing the signature to run() back then!

**Comment by mathmarques** [comment]
@akrabat Wouldn't be good to call `createResponse()` instead of this:
https://github.com/akrabat/Slim/blob/deprecate-request-response-env-from-container/Slim/App.php#L501-L503

**Comment by akrabat** [comment]
Yes - I've updated.

File: Slim/DefaultServicesProvider.php [hunk 1]
```diff
@@ -37,47 +37,6 @@ class DefaultServicesProvider
...
-             *
-             * @return ResponseInterface
-             */
-            $container['response'] = function ($container) {
-                $headers = new Headers(['Content-Type' => 'text/html; charset=UTF-8']);
-                $response = new Response(200, $headers);
-
-                return $response->withProtocolVersion($container->get('settings')['httpVersion']);
-            };
-        }
```
Comment (mnapoli) [inline Slim/DefaultServicesProvider.php:79]: Why not leave these here but not actually use them in `createRequest()` and `createResponse()`? That way you keep BC, you can tag these entries as deprecated and `App` is much simpler.

File: Slim/App.php [hunk 2]
```diff
@@ -481,7 +488,7 @@ public function subRequest(
         $bodyContent = '',
         ResponseInterface $response = null
     ) {
-        $env = $this->container->get('environment');
+        $env = new Environment($_SERVER);
```
Comment (mnapoli) [inline Slim/App.php:491]: Do you think it would make sense to have something like `Environment::fromGlobal()` that would do the same thing?

File: Slim/DefaultServicesProvider.php (hunk 1 above)
Comment (akrabat) [inline Slim/DefaultServicesProvider.php:79]: If someone has added their own request/response to the container before calling run() they will expect that their container will be used.  The easiest way to tell that this has happened is to remove our default implementations.

File: Slim/App.php (hunk 2 above)
Comment (akrabat) [inline Slim/App.php:491]: Maybe, but that would be a different PR.

File: Slim/DefaultServicesProvider.php (hunk 1 above)
Comment (mnapoli) [inline Slim/DefaultServicesProvider.php:79]: Right sorry that was probably what you meant before but I didn't understand it that way.

File: Slim/App.php (hunk 2 above)
Comment (mnapoli) [inline Slim/App.php:491]: :+1:

File: Slim/App.php [hunk 3]
```diff
@@ -630,4 +637,91 @@ protected function handlePhpError(Throwable $e, ServerRequestInterface $request,
...
+        }
+
+        if ($this->container->has('environment')) {
+            $environment = $this->container->get('environment');
+        } else {
+            $environment = new Environment($_SERVER);
+
+            // register our new environment with the container for BC
+            if ($this->container instanceof \ArrayAccess) {
+                if (!$this->container->has('environment')) {
```
Comment (vlakoff) [inline Slim/App.php]: Do we need this test? The `else` line 657 tells us the container doesn't have environment.

Comment (akrabat) [inline Slim/App.php]: Good point. Removed.

**Comment by juliangut** [comment]
How would this affect slim/twig-view `TwigExtension`? It extracts the request from container to retrieve basePath, how would this be done instead?

``` php
// Register Twig View helper
$container['view'] = function ($c) {
    $view = new \Slim\Views\Twig('path/to/templates', [
        'cache' => 'path/to/cache'
    ]);

    // Instantiate and add Slim specific extension
    $basePath = rtrim(str_ireplace('index.php', '', $c['request']->getUri()->getBasePath()), '/');
    $view->addExtension(new Slim\Views\TwigExtension($c['router'], $basePath));

    return $view;
};
```

**Comment by mathmarques** [comment]
We still be causing a BC break. If someone try to get request/response from container before call `run`.
Good catch @juliangut, but in that case it won't generate error unless you use view before call `run`(no problem using inside middleware/route callback).

**Comment by akrabat** [comment]
Still more thinking required!

**Comment by geggleto** [comment]
I love this.
Yes plz.
:+1:

**Comment by geggleto** [comment]
Should we not put into the container actual Factories for Request/Response/Environment?

File: Slim/App.php (hunk 2 above)
Comment (geggleto) [inline Slim/App.php:491]: @akrabat this should be put elsewhere so we can test it properly.

File: Slim/App.php [hunk 4]
```diff
@@ -288,10 +290,15 @@ public function group($pattern, $callable)
      * @throws MethodNotAllowedException
      * @throws NotFoundException
      */
-    public function run($silent = false)
+    public function run($silent = false, ServerRequestInterface $request = null, ResponseInterface $response = null)
     {
-        $request = $this->container->get('request');
-        $response = $this->container->get('response');
+        if (!$request) {
+            $request = $this->createRequest();
```
Comment (geggleto) [inline Slim/App.php:296]: I really think these should be put on the actual Http Object.

**Comment by akrabat** [comment]
> How would this affect slim/twig-view TwigExtension? It extracts the request from container to retrieve basePath, how would this be done instead?

It won't affect it. This PR registers request and response with the container so that when the view is retrieved from the container, the objects are there.

However, we need a new pattern as a deprecation notice will be raised.

**Comment by juliangut** [comment]
> However, we need a new pattern as a deprecation notice will be raised.

That was what I was refering to :smile:, I see a deprecation notice when there was none as a way of affecting behavior

**Comment by akrabat** [comment]
Needs to be done in 4.0.
